oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/web/authentication/logout/OidcBackChannelLogoutFilter.java

```diff
@@ -0,0 +1,205 @@
+import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+/**
+ * A filter for the Client-side OIDC Backchannel Logout endpoint
+ *
+ * @author Josh Cummings
+ * @since 6.1
+ * @see <a target="_blank" href=
+ * "https://openid.net/specs/openid-connect-backchannel-1_0.html">OIDC Backchannel Logout
+ * Spec</a>
+ */
+public class OidcBackChannelLogoutFilter extends OncePerRequestFilter {
+
+	private final Log logger = LogFactory.getLog(getClass());
+
+	private static final String ERROR_MESSAGE = "{ \"error\" : \"%s\", \"error_description\" : \"%s\" }";
+
+	private final ClientRegistrationRepository clientRegistrationRepository;
+
+	private final AuthenticationManager authenticationManager;
+
+	private RequestMatcher requestMatcher = new AntPathRequestMatcher("/logout/connect/back-channel/{registrationId}",
+			"POST");
+
+	private OidcProviderSessionRegistry providerSessionRegistry = new InMemoryOidcProviderSessionRegistry();
+
+	private LogoutHandler logoutHandler = new BackchannelLogoutHandler();
+
+	/**
+	 * Construct an {@link OidcBackChannelLogoutFilter}
+	 * @param clients the {@link ClientRegistrationRepository} for deriving Logout Token
+	 * validation
+	 * @param logoutTokenDecoderFactory the {@link JwtDecoderFactory} for constructing
+	 * Logout Token validators
+	 */
+	public OidcBackChannelLogoutFilter(ClientRegistrationRepository clients,
+			AuthenticationManager authenticationManager) {
+		this.clientRegistrationRepository = clients;
+		this.authenticationManager = authenticationManager;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+			throws ServletException, IOException {
+		RequestMatcher.MatchResult result = this.requestMatcher.matcher(request);
+		if (!result.isMatch()) {
+			chain.doFilter(request, response);
+			return;
+		}
+		String registrationId = result.getVariables().get("registrationId");
+		ClientRegistration registration = this.clientRegistrationRepository.findByRegistrationId(registrationId);
+		if (registration == null) {
+			this.logger.debug("Did not process OIDC Backchannel Logout since no ClientRegistration was found");
+			chain.doFilter(request, response);
+			return;
+		}
+		String logoutToken = request.getParameter("logout_token");
+		if (logoutToken == null) {
+			String error = "Failed to process OIDC Backchannel Logout since no logout token was found";
+			this.logger.debug(error);
+			String message = String.format(ERROR_MESSAGE, "invalid_request", error);
+			response.sendError(400, message);
+			return;
+		}
+		OidcLogoutToken token;
+		try {
+			token = authenticate(logoutToken, registration);
+		}
+		catch (AuthenticationException ex) {
```

Comment:
That would mean that any `AuthenticationException` would result in a 500, at least in this case; I think we should be more robust than that. To that end, I'll update it to catch `AuthenticationServiceException` as well, in order to send a 500 in those cases.

Note that my review of `OAuth2LoginAuthenticationFilter` is that it catches `AuthenticationException` and `InternalAuthenticationServiceException` in the `doFilter` method that it inherits from `AbstractAuthenticationProcessingFilter`. Did I miss something?

Reply:
@jzheaux 

> Note that my review of `OAuth2LoginAuthenticationFilter` is that it catches `AuthenticationException` and `InternalAuthenticationServiceException`

I must be missing something because I don't see the same. Can you please provide a link to the code that you reviewed?

Reply:
https://github.com/spring-projects/spring-security/blob/main/web/src/main/java/org/springframework/security/web/authentication/AbstractAuthenticationProcessingFilter.java#L247

Reply:
Based on a private discussion I had with Joe, I'd like to at this point try and clear up some misunderstanding.

It seems like the core difference of opinion here is what to do when it is a server-side error. An example of this is if a configured `AuthenticationManager` makes a remote call that fails due to the remote service being down. If we follow HTTP semantics, then [a 500 is appropriate in this case](https://www.rfc-editor.org/rfc/rfc9110.html#status.5xx):

> The 500 (Internal Server Error) status code indicates that the server encountered an unexpected condition that prevented it from fulfilling the request.

It has been argued that because the backchannel logout spec states the following:

> If the logout request was invalid or the logout failed, the RP MUST respond with HTTP 400 Bad Request

that this means that anything that would conventionally be a 500 error should instead be adapted to a 400 error.

This being the core difference of opinion, I'd like to offer why it's clear to me that it's not the intent of the spec to have 500 errors be overridden and that indeed sending them in response to a backchannel request is valuable for authorization servers.

First, there is already room in the spec based on this statement at the beginning:

> This specification is designed for use with HTTP ([[RFC2616](https://datatracker.ietf.org/doc/html/rfc2616)])

Second, the back-channel spec points to section 5.2 of the OAuth 2.0 spec for error codes: 

> These error response parameters are used as specified in Section 5.2 of [OAuth 2.0](https://openid.net/specs/openid-connect-backchannel-1_0.html#RFC6749) [RFC6749]

These [error codes](https://datatracker.ietf.org/doc/html/rfc6749#section-5.2) are:

```
invalid_request
invalid_client
invalid_grant
unauthorized_client
unsupported_grant_type
invalid_scope
```

none of which applies in the case of a server-side error. IOW, these are all by name and by definition client-side problems that the client can address.

Third, the spec states its purpose as (emphasis **mine**):

> It enables Clients to verify the identity of the End-User based on the authentication performed by an Authorization Server, as well as to obtain basic profile information about the End-User **in an interoperable and REST-like manner**. 

which underscores that the intent of the spec is to not get in the way of REST semantics.

It's true that the OAuth 2.0 specification defines `server_error`. However, it defines this *only* in the Authorization Code Grant section. Also, it gives a specific caveat for the reason for its existence (emphasis **mine**):

> The authorization server encountered an unexpected condition that prevented it from fulfilling the request.
> (**This error code is needed because a 500 Internal Server Error HTTP status code cannot be returned to the client via an HTTP redirect.**)

Since OIDC back-channel logout is not a redirect-based protocol, it does not make the same carve out for `server_error`. Further, why add a special description here for 400 over redirect if everyone is assuming that 500s are globally to be mapped as 400s? Indeed, I did a bit of digging about this [in the mailing list](https://mailarchive.ietf.org/arch/msg/oauth/dz7KDpwg-xS6b38l9Dpdc_5Fic0/) and found the following (emphasis **mine**): 

> But they're not needed for non-implicit grant types **because you've got 
actual HTTP error codes for those things, then.**
> 
> I don't think this is a valid errata.

This was in response to a request to edit Section 5.2 to include `server_error` (please see the link for full details). There is also a related discussion about [why `server_error` was added](https://mailarchive.ietf.org/arch/msg/oauth/Th_876-Iqf7yZKmwVAnbGYZUifQ/) specifically because 500 wasn't an option in that specific circumstance; otherwise, 500 would have been appropriate.

Next, the most practical assumption is that 500s are going to happen. They happen when there is an accidental `NullPointerException` or some other programmer error. Since 500s can then already happen (we cannot prevent them), we should instead align with the semantics of those circumstances. That is, when it is the server's fault (NPE, misconfig, server down), we say so. This gives the client important pragmatic, actionable information that the client will need to handle anyway. Why make them look for 500s AND look for 400 `server_error` to determine a server-side error? We should keep it simple for the caller and just ask them to listen for 500s to indicate a server-side error.

Finally, as I've already linked to, all filters that extend from `AbstractAuthenticationProcessingFilter` already do exactly what I've just outlined. If an `AuthenticationServiceException` is thrown, these filters handle this exception differently from `AuthenticationException`. This is also true with back-end filters like `BearerTokenAuthenticationFilter` and the default settings for `AuthenticationFilter` and `AuthenticationWebFilter` -- they use `AuthenticationEntryPointFailureHandler` which by default induces a 500 error for `AuthenticationServiceException`s.

So, here's what I propose changing. The filter should change to:

```java
try {
    // ... do work
} catch (OAuth2AuthenticationException client) {
   // ... respond with 400, including any OAuth2Error as detail
} catch (AuthenticationServiceException server) {
   // ... respond with 500 since this indicates a server-side error
} catch (AuthenticationException client) {
   // ... respond with generic 400, according to spec
}
```
This is so that any `OAuth2Error` that the `AuthenticationManager` wants to return can be transmitted. While I would vote for `invalid_token` as I feel that's the clearest indication of what happened, if we are to adhere to the spec, the closest is `invalid_request`.

The reason for catching `AuthenticationServiceException` is because this is a generic exception that "something went wrong" and really has nothing, per se, to do with OAuth or OIDC. Also, an implementer of `AuthenticationManager` may throw an `AuthenticationException` and we should not break in those circumstances. This codes nicely to the `AuthenticationManager` contract which says:

```
Authentication authenticate(Authentication authentication) throws AuthenticationException;
```

IOW, because those who implement `AuthenticationManager` *may* throw an `AuthenticationException` or an `AuthenticationServiceException`, we should robustly handle both.

Also, `OidcBackChannelLogoutAuthenticationManager` should change to:

```java
} catch (BadJwtException ex) {
    throw new OAuth2AuthenticationException(...);
} catch (Exception ex) {
    throw new AuthenticationServiceException(ex.getMessage(), ex);
}
```

This allows the authentication manager to give an `OAuth2Error` code. It also ensures that server-side errors are propagated to the filter. Note that while this is different than the OAuth 2.0 authentication providers, it is consistent with at least `DaoAuthenticationProvider`, `JwtAuthenticationProvider`, and `OpaqueTokenAuthenticationProvider`. I didn't perform an exhaustive search; I'm certain the codebase isn't consistent, but this is the direction it should head.

I do not support the creation of an `OAuth2AuthenticationServiceException` (an idea Joe and I tossed around privately) since it is unneeded. There is no `OAuth2Error` code that the spec indicates for back-end server-side errors and so `AuthenticationServiceException` contains all the information needed to transmit back to the client and still adhere to the spec.

Reply:
Thanks for taking the time to provide such detailed information @jzheaux.

After reviewing RFC 6749, I agree that `server_error` should not be used as the `error` parameter in the back-channel flow. As the spec states, it's meant to be returned in redirect-based flows, which is not applicable here.

Regarding

> Second, the back-channel spec points to section 5.2 of the OAuth 2.0 spec for error codes:

Section 5.2 in RFC 6749 is specific to the Error Response following a Token Request so it's not applicable here.

Here is my interpretation of [2.8.  Back-Channel Logout Response](https://openid.net/specs/openid-connect-backchannel-1_0.html#BCResponse)

> If the logout request was invalid or the logout failed, the RP MUST respond with HTTP 400 Bad Request. The response MAY include an HTTP body consisting of a JSON object with error and error_description parameters conveying the nature of the error that occured, which can assist with debugging. These error response parameters are used as specified in Section 5.2 of [OAuth 2.0](https://openid.net/specs/openid-connect-backchannel-1_0.html#RFC6749) [RFC6749]

> If the logout request was invalid or the logout failed, the RP MUST respond with HTTP 400 Bad Request.

Since **MUST** is used, this is a strict requirement to return 400 as the status, since the AS will expect this status code.

> The response MAY include an HTTP body consisting of a JSON object with error and error_description parameters conveying the nature of the error that occured, which can assist with debugging.

Since **MAY** is used, an OAuth2Error JSON is optional but since it can assist in debugging, it is recommended.

> These error response parameters are used as specified in Section 5.2 of [OAuth 2.0](https://openid.net/specs/openid-connect-backchannel-1_0.html#RFC6749) [RFC6749]

I interpret this if an OAuth2Error JSON is returned it should follow the format as used in the Error Response specified in Section 5.2, however, it does not state that one of those error codes should be used.

Returning a 500 status is not spec compliant.

As stated in 2.6.  Logout Token Validation:

> If any of the validation steps fails, reject the Logout Token and return an HTTP 400 Bad Request error. Otherwise, proceed to perform the logout actions.

and in 2.8.  Back-Channel Logout Response:

> If the logout request was invalid or the logout failed, the RP MUST respond with HTTP 400 Bad Request.

The wording MUST gives us no choice but to return 400, otherwise, we are introducing a bug from the get go.

I took a look at `OidcAuthorizationCodeAuthenticationProvider` to see how it handles the remote call server exception. Take a look at this line

https://github.com/spring-projects/spring-security/blob/2ef9dc916fe5d98a99319c72f575b1f5c8cdf474/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java#L247

You will notice it handles `JwtException` only and then if you look at this line in `NimbusJwtDecoder`:

https://github.com/spring-projects/spring-security/blob/2ef9dc916fe5d98a99319c72f575b1f5c8cdf474/oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java#L170

A `JwtException` is thrown with the nested `RemoteKeySourceException`, which is the exception that is thrown during a remote service call to the client JWKSet endpoint.

Based on this, I would recommend changing the logic in `OidcBackChannelLogoutAuthenticationProvider` to catch `JwtException` and throw an `OAuth2AuthenticationException` with the nested `JwtException`. The `OidcBackChannelLogoutFilter` could potentially use the `JwtException.getMessage` as the `error_description` and we can either return `invalid_request` as the `error` parameter or return a custom `error` that we define similar to how `OidcAuthorizationCodeAuthenticationProvider` returns a custom `error` as `invalid_id_token`.

Reply:
I think this is too narrow of an interpretation of the spec as it doesn't take the broader context into account.

1. The overall language of the spec indicates that it's intended to be compatible with REST semantics; the mailing list comments further back this up. While I don't want this to turn into an appeal of authority, if needed, I also have word from one of the spec writers that returning 500s is fine. On these grounds, I disagree that the spec mandates masking 500s as 400s.
2. Masking 500s makes it tricker for clients since they must now watch for 500 as well as 400 errors and now try and interpolate from a 400 whether it is really a 500.

I don't see the spec asking us to make things more complicated for clients so I don't know why we should.

Reply:
> I also have word from one of the spec writers that returning 500s is fine

In the general case, I agree this would be fine. However, no where in the spec does it mention returning 500 in the flow. It clearly indicates that 400 MUST be returned.

Either way, I'm looking to move forward here and we'll simply agree to disagree which is totally fine. Please go ahead and return 500 when the remote call to client's JWKSet fails. I'm not sure how you plan on handling this but consider my previous comment:

> I took a look at `OidcAuthorizationCodeAuthenticationProvider` to see how it handles the remote call server exception. Take a look at this line
>
> https://github.com/spring-projects/spring-security/blob/2ef9dc916fe5d98a99319c72f575b1f5c8cdf474/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java#L247
>
> You will notice it handles `JwtException` only and then if you look at this line in `NimbusJwtDecoder`:
>
> https://github.com/spring-projects/spring-security/blob/2ef9dc916fe5d98a99319c72f575b1f5c8cdf474/oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java#L170
>
> A `JwtException` is thrown with the nested `RemoteKeySourceException`, which is the exception that is thrown during a remote service call to the client JWKSet endpoint.

It appears the only way you can handle this is by inspecting the thrown `JwtException` for the nested exception of type `RemoteKeySourceException`. This may work for now, but if a custom `JwtDecoder` is used and is not using Nimbus implementation then that failure scenario will force a 400 return because there won't be a nested `RemoteKeySourceException`. Anyway, something to consider when you apply the updates.

Reply:
> However, no where in the spec does it mention returning 500 in the flow.

The spec doesn't need to say this other than to say that it's built on top of HTTP and REST semantics (which it does). The appropriate response codes can be inferred from lower levels of the protocol stack.

> Please go ahead and return 500 when the remote call to client's JWKSet fails

This is unnecessarily narrow (there are all kinds of things that are not the client's fault). We can make the code simpler by catching `Exception` and wrapping it in an `AuthenticationServiceException`. This is Spring Security's built-in signal that something unattributable to the client happened during authentication.

In other words, the most pragmatic orientation towards errors is this: If we can determine it's the client's fault, then we say so. If we can't then, it's the server's fault. This is a common default stance due to its simplicity. It also allows us to improve over time, returning more 400s as we get clearer on what is causing various problems.

Thus, my recommendation remains as follows:

```java
} catch (BadJwtException ex) { // we know what went wrong
    throw new OAuth2AuthenticationException(...);
} catch (Exception ex) { // we don't know what went wrong
    throw new AuthenticationServiceException(...);
}
```
This is a nice stance since it allows us to wrap other unexpected behaviors like programmer errors and have a consistent outward behavior. 

As we improve our abilities over time to discern problems, we can add more exceptions and give more information to ppl:

```java
} catch (BadJwtException ex) { // we know what went wrong
    throw new OAuth2AuthenticationException(...);
} catch (AnotherKindOfClientErrorException ex) { // we know what went wrong
    throw new OAuth2AuthenticationException(...);
} catch (Exception ex) { // we don't know what went wrong
    throw new AuthenticationServiceException(...);
}
```

Reply:
> Thus, my recommendation remains as follows:

```java
} catch (BadJwtException ex) { // we know what the client did wrong
    throw new OAuth2AuthenticationException(...);
} catch (Exception ex) { // we don't know what went wrong
    throw new AuthenticationServiceException(...);
}
```

Ok. Let's go with this.